File: android-beans/src/main/java/rocks/frieler/android/beans/SingleBeanDependency.java
package rocks.frieler.android.beans;

/**
 * {@link BeanDependency} to express a dependency on a single bean.
 *
 * @param <T> the bean-type
 */
public class SingleBeanDependency<T> implements BeanDependency<T> {
    private final String name;
    private final Class<? extends T> type;
    private T bean = null;


    /**
     * Creates a new {@link BeanDependency} to a bean of the given type.
     *
     * @param type the type of the bean
     */
    public SingleBeanDependency(Class<? extends T> type) {
        this(null, type);
    }

    /**
     * Creates a new {@link BeanDependency} to a bean with the given name and of the given type.
     *
     * @param name the name of the bean
     * @param type the type of the bean
     */
    public SingleBeanDependency(String name, Class<? extends T> type) {
        this.name = name;
        this.type = type;
    }

    @Override
    public Fulfillment fulfill(BeansProvider beansProvider) {
        if (bean == null) {
            bean = (name == null ? beansProvider.lookUpBean(type) : beansProvider.lookUpBean(name, type));
        }
        return (bean != null ? Fulfillment.FULFILLED : Fulfillment.UNFULFILLED);
    }

    @Override
    public T get() {
        return bean;
    }
}

File: android-beans/src/main/kotlin/rocks/frieler/android/beans/SingleBeanDependency.kt
package rocks.frieler.android.beans

import rocks.frieler.android.beans.BeanDependency.Fulfillment
import java.util.Objects
import kotlin.reflect.KClass

/**
 * [BeanDependency] to express a dependency on a single bean.
 *
 * @param <T> the bean-type
 */
class SingleBeanDependency<T :Any>(
		private val name: String?,
		private val type: KClass<out T>
) : BeanDependency<T> {

	constructor(type: KClass<out T>) : this(null, type)

	private var bean: T? = null

	override fun fulfill(beansProvider: BeansProvider): Fulfillment {
		if (bean == null) {
			bean = if (name == null) beansProvider.lookUpBean(type) else beansProvider.lookUpBean(name, type)
		}
		return if (bean != null) Fulfillment.FULFILLED else Fulfillment.UNFULFILLED
	}

	override fun equals(other: Any?): Boolean {
		return if (javaClass == other?.javaClass) {
			other as SingleBeanDependency<*>
			Objects.equals(name, other.name) && Objects.equals(type, other.type)
		} else {
			false
		}
	}

	override fun hashCode(): Int {
		return Objects.hash(name, type)
	}
}
